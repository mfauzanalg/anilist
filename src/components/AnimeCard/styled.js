import styled from "@emotion/styled";

export const AnimeCardContainer = styled.div`
  width: 150px;
  margin: 5px 8px;
  position: relative;
`

export const AnimePosterContainer = styled.div`
  width: 100%;
  height: 250px;
  overflow: hidden;
  border-radius: 10px;
`

export const AnimePosterImg = styled.img`
  width: 100%;
  height: 280px;
  border-radius: 10px;
  object-fit: cover;
`

export const AnimeTitle = styled.div`
  color: white;
  font-size: 14px;
  margin-top: 5px;
  position: absolute;
  bottom: 10px;
  left: 10px;
  width: 90%;
  border: 1px solid transparent;
  font-weight: bold;
  opacity: 0.9;
  z-index: 99;
`

export const BottomShadow = styled.div`
  position: absolute;
  bottom: -3px;
  z-index: 9;
  width: 100%;
  height: 40%;
  background-image: linear-gradient(transparent, #141414);
`