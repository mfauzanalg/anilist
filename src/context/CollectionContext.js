import { useState, createContext, useEffect } from 'react';

export const CollectionContext = createContext();
export const CollectionProvider = (props) => {

  const [collections, setCollections] = useState(JSON.parse(localStorage.getItem('collections')) || [])

  useEffect(() => {
    localStorage.setItem('collections', JSON.stringify(collections))
  }, [collections])

  const addNewCollection = (name) => {
    const temp = [...collections]
    temp.push({ name: name })

    setCollections(temp);
  }

  return (
    <CollectionContext.Provider value={{ collections, addNewCollection }}>
      {props.children}
    </CollectionContext.Provider>
  )
}