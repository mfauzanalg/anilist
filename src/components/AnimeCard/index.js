import { AnimeCardContainer, AnimePosterContainer, AnimePosterImg, AnimeTitle, BottomShadow, CardContainer, RemoveContainer } from './styled'
import { useNavigate } from 'react-router-dom';
import ConfirmationModal from '../Modal/ConfirmationModal';
import { useContext, useState } from 'react';
import { CollectionContext } from '../../context/CollectionContext';
import CornerButton from '../CornerButton';
import CloseIcon from '@mui/icons-material/Close';
import toast from 'react-hot-toast';

const AnimeCard = ({ title, coverImage, id, isHideDelete, collectionName }) => {
  const navigate = useNavigate()
  const { removeAnimeFromCollection } = useContext(CollectionContext)
  const [isOpenDialog, setIsOpenDialog] = useState(false);

  const handleOnClick = () => {
    navigate(`/anime/${id}`)
  }

  const handleRemoveMondal = () => {
    setIsOpenDialog(true)
  }

  const handleRemoveAnime = () => {
    removeAnimeFromCollection(id, collectionName)
    setIsOpenDialog(false)
    toast.success('Success remove anime')
  }

  return (
    <AnimeCardContainer isHideDelete={isHideDelete}>
      <CardContainer onClick={handleOnClick}>
        <AnimePosterContainer>
          <AnimePosterImg src={coverImage} alt='coverImage' />
        </AnimePosterContainer>
        <AnimeTitle>
          {title}
        </AnimeTitle>
        <BottomShadow />
      </CardContainer>

      {!isHideDelete &&
        <RemoveContainer onClick={handleRemoveMondal}>
          <CornerButton color={'gray'} size={'20px'}>
            <CloseIcon fontSize='10px' />
          </CornerButton>
        </RemoveContainer>
      }

      {/* Modals */}
      <ConfirmationModal
        title={'Remove from Collection'}
        text={`Are you sure want to remove ${title}?`}
        open={isOpenDialog}
        setOpen={setIsOpenDialog}
        primaryButtonText={'Remove'}
        onPrimaryClick={handleRemoveAnime}
        onSecondaryClick={() => { setIsOpenDialog(false) }}
      />

    </AnimeCardContainer>
  )
}

export default AnimeCard