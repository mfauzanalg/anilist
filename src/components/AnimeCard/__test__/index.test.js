import { render, screen } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect';
import { MemoryRouter } from 'react-router-dom';
import { CollectionProvider } from '../../../context/CollectionContext';
import AnimeCard from '..'
import renderer from "react-test-renderer";

const animeCardAttribute = {
  title: 'Anime title',
  coverImage: 'https://s4.anilist.co/file/anilistcdn/media/anime/cover/medium/bx1-CXtrrkMpJ8Zq.png',
  id: 1,
  isHideDelete: false
}

const animeCardAttributeHideDelete = {
  title: 'Anime title',
  coverImage: 'https://s4.anilist.co/file/anilistcdn/media/anime/cover/medium/bx1-CXtrrkMpJ8Zq.png',
  id: 1,
  isHideDelete: true
}

const renderComponent = (attribute) => {
  return render(
    <CollectionProvider>
      <MemoryRouter>
        <AnimeCard {...attribute} />
      </MemoryRouter>
    </CollectionProvider>
  )
}

it('AnimeCard rendered correctly', () => {
  renderComponent(animeCardAttribute);
})

it('AnimeCard has correct name', () => {
  renderComponent(animeCardAttribute);
  expect(screen.getByTestId('anime-card-title')).toHaveTextContent('Anime title');
})

it('AnimeCard with close button', () => {
  renderComponent(animeCardAttribute);
  expect(screen.getByRole('button', { color: 'gray' })).toBeInTheDocument();
  expect(screen.getByRole('button', { color: 'gray' })).toBeEnabled();
})

it('AnimeCard without close button', () => {
  renderComponent(animeCardAttributeHideDelete);
  expect(screen.queryByTestId('corner-button')).toBeNull();
})

it("AnimeCard matches snapshot", () => {
  const tree = renderer.create(
    <CollectionProvider>
      <MemoryRouter>
        <AnimeCard {...animeCardAttribute} />
      </MemoryRouter>
    </CollectionProvider>
  ).toJSON;
  expect(tree).toMatchSnapshot();
});
