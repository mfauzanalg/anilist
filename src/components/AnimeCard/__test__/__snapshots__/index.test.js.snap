// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnimeCard matches snapshot 1`] = `[Function]`;

exports[`Card matches snapshot 1`] = `[Function]`;
